src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java

```diff
@@ -275,107 +422,164 @@ private String getClassName(JarEntry f) {
      * @param name name
      * @param predicate predicate
      * @return true if yes
+     *
+     * @see RuntimeEnvironment#getPluginConfiguration()
      */
     @SuppressWarnings("unchecked")
-    private boolean checkAll(HttpServletRequest request, String cache, String name,
-            Predicate<IAuthorizationPlugin> predicate) {
+    private boolean checkAll(HttpServletRequest request, String cache, Nameable entity,
+            Predicate<AuthorizationPluginWrapper> predicate) {
+        Boolean val;
         Map<String, Boolean> m = (Map<String, Boolean>) request.getAttribute(cache);
 
         if (m == null) {
             m = new TreeMap<>();
+        } else if ((val = m.get(entity.getName())) != null) {
+            // cache hit
+            return val;
         }
 
-        Boolean val = m.get(name);
+        boolean overallDecision = performCheck(entity, predicate);
 
-        if (val != null) {
-            return val;
-        }
+        m.put(entity.getName(), overallDecision);
+        request.setAttribute(cache, m);
+        return overallDecision;
+    }
 
-        for (IAuthorizationPlugin plugin : getPlugins()) {
+    private boolean performCheck(Nameable entity, Predicate<AuthorizationPluginWrapper> predicate) {
+        boolean overallDecision = true;
+        for (AuthorizationPluginWrapper plugin : getPlugins()) {
+            // run the plugin's test method
             try {
-                if (!predicate.test(plugin)) {
-                    m.put(name, false);
-                    request.setAttribute(cache, m);
-                    return false;
+                boolean pluginDecision = predicate.test(plugin);
+
+                if (!pluginDecision && plugin.isRequired()) {
+                    // required sets a failure but still invokes all other plugins
+                    overallDecision = false;
+                    continue;
+                } else if (!pluginDecision && plugin.isRequisite()) {
+                    // requisite sets a failure and immediately returns the failure
+                    overallDecision = false;
+                    break;
+                } else if (overallDecision && pluginDecision && plugin.isSufficient()) {
+                    // sufficient immediately returns the success
+                    overallDecision = true;
+                    break;
                 }
             } catch (Throwable ex) {
-                LOGGER.log(Level.SEVERE, "Plugin \"" + plugin.getClass().getName() + "\" has failed with exception:", ex);
-                removePlugin(plugin);
+                LOGGER.log(Level.WARNING,
+                        String.format("Plugin \"%s\" has failed the testing of \"%s\" with an exception.",
+                                plugin.getClassname(),
+                                entity.getName()),
+                        ex);
+
+                // set the return value to false for this faulty plugin
+                if (!plugin.isSufficient()) {
```

Comment:
if SUFFICIENT plugin throws exception, it should return failure too I think. In overall I wonder if any exception here should cause immediate termination of traversal and return false.